plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/repl/ReplCodeAnalyzer.kt

```diff
@@ -265,11 +265,18 @@ open class ReplCodeAnalyzerBase(
         }
 
         private fun computeFileScopes(lineInfo: LineInfo, fileScopeFactory: FileScopeFactory): FileScopes? {
+            val linePsi = lineInfo.linePsi
+            val hasImports = linePsi.importDirectives.isNotEmpty()
+
             // create scope that wraps previous line lexical scope and adds imports from this line
             val lexicalScopeAfterLastLine = lineInfo.parentLine?.lineDescriptor?.scopeForInitializerResolution ?: return null
             val lastLineImports = lexicalScopeAfterLastLine.parentsWithSelf.first { it is ImportingScope } as ImportingScope
-            val scopesForThisLine = fileScopeFactory.createScopesForFile(lineInfo.linePsi, lastLineImports)
-            val combinedLexicalScopes = lexicalScopeAfterLastLine.replaceImportingScopes(scopesForThisLine.importingScope)
+            val scopesForThisLine = fileScopeFactory.createScopesForFile(linePsi, lastLineImports)
+            val combinedLexicalScopes = if (hasImports)
+                lexicalScopeAfterLastLine.replaceImportingScopes(scopesForThisLine.importingScope)
+            else
+                lexicalScopeAfterLastLine
```

Comment:
Have you checked that this does not affect implicit imports, that are passed via script compilation configuration?
I believe it shouldn't, but it must be checked, I believe.
You'll need to construct a REPL with a custom script def with implicit imports, to check that. Maybe it makes sense to add tests for it too, we have not many tests with imports and this change can alter the behavior.

Reply:
@ligee are you talking about `defaultImports`, right?

Reply:
If so, I've a test